src/GafferScene/Preview/InteractiveRender.cpp

```diff
@@ -131,55 +144,188 @@ class InteractiveRender::SceneGraph
 			return m_name;
 		}
 
-		// Returns false if the attributes make the location invisible, true otherwise.
-		bool updateAttributes( const CompoundObjectPlug *attributesPlug, IECoreScenePreview::Renderer *renderer )
+		// Called by SceneGraphUpdateTask to update this location. Returns a bitmask
+		// of the components which were changed.
+		unsigned update( const ScenePlug *scene, unsigned dirtyComponents, unsigned changedParentComponents, Type type, IECoreScenePreview::Renderer *renderer, const IECore::CompoundObject *globals )
 		{
-			const IECore::MurmurHash attributesHash = attributesPlug->hash();
-			if( attributesHash == m_attributesHash && !parentPending( AttributesPending ) )
+			unsigned changedComponents = 0;
+
+			// Attributes
+
+			if( !m_parent )
+			{
+				// Root - get attributes from globals.
+				if( dirtyComponents & GlobalsComponent )
+				{
+					if( updateAttributes( globals ) )
+					{
+						changedComponents |= AttributesComponent;
+					}
+				}
+			}
+			else
 			{
-				return true;
+				// Non-root - get attributes the standard way.
+				const bool parentAttributesChanged = changedParentComponents & AttributesComponent;
+				if( parentAttributesChanged || ( dirtyComponents & AttributesComponent ) )
+				{
+					if( updateAttributes( scene->attributesPlug(), parentAttributesChanged ) )
+					{
+						changedComponents |= AttributesComponent;
+					}
+				}
 			}
 
-			ConstCompoundObjectPtr attributes = attributesPlug->getValue( &attributesHash );
-			CompoundObject::ObjectMap &fullAttributes = m_fullAttributes->members();
-			if( m_parent )
+			if( !::visible( m_fullAttributes.get() ) )
 			{
-				fullAttributes = m_parent->m_fullAttributes->members();
+				clear();
+				return changedComponents;
 			}
-			else
+
+			// Transform
+
+			if( ( dirtyComponents & TransformComponent ) && updateTransform( scene->transformPlug(), changedParentComponents & TransformComponent ) )
+			{
+				changedComponents |= TransformComponent;
+			}
+
+			// Object
+
+			if( ( dirtyComponents & ObjectComponent ) && updateObject( scene->objectPlug(), type, renderer, globals ) )
+			{
+				changedComponents |= ObjectComponent;
+			}
+
+			// Object updates for transform and attributes
+
+			if( m_objectInterface )
+			{
+				if( changedComponents & ObjectComponent )
+				{
+					// New object always needs transform applying.
```

Comment:
does the new object always need attributes applying to? or did that happen above already?